### PR TITLE
fixed torch.where and torch.nonzero issue

### PR DESCRIPTION
**Pull request by alpoge**
After careful search it seems mpsGraph's scatter method inadvertently casts an updatesTensor which contains int32s to float32, which leads to rounding for integers larger than 2^(24) due to floating point representation. This has led to errors like #122233 and #122916 , which concern torch.where (really torch.nonzero) calls which give an incorrect answer (explained by the above: it turns out the incorrect output is the result of casting the correct output to float32 and then back to int32...). I am not sure to what extent it is responsible for the other MPS arithmetic issues, but I will investigate. My solution is the straightforward one: split the coordinatesTensor into two tensors with entries which won't be incorrectly rounded by the float32 cast, scatter them both, then recombine.

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/Indexing.mm [hunk 1]
```diff
@@ -369,13 +369,45 @@ static Tensor nonzero_fallback(const Tensor& self) {
         maskedIndicesTensor = [mpsGraph concatTensors:maskedIndicesTensorArray dimension:0 interleave:YES name:nil];
         coordinatesTensor = [mpsGraph concatTensors:coordinatesTensorArray dimension:0 interleave:YES name:nil];
       }
-
-      MPSGraphTensor* outputTensor = [mpsGraph scatterWithDataTensor:scatterDataTensor
+    
+    MPSGraphTensor* outputTensor;
+
+    // mpsGraph's scatter seems to be, as of this commit, casting int32's (like coordinatesTensor) to float32's and then back,
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/Indexing.mm:375]: This was indeed the case. We are currently doing intermediate indices through F32 conversion which is reducing the range of indexing. We are looking into more efficient way of addressing this issue.

Comment (kulinseth) [reply aten/src/ATen/native/mps/operations/Indexing.mm:375]: Overall the spirit of PR is great.. Thanks for the fix

**Comment by kulinseth** [comment]
Can you please add a test to `test_mps` for this issue in your PR , so that we don't regress this in the future.

**Comment by mahdiaslanimk** [comment]
I had somewhat the same issue as two reported ones above.
Code to reproduce the issue:
```python
mytens = torch.tensor([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                        [1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 1., 0.]], device='mps')
                        
for i in range(10):
    print(torch.nonzero(mytens, as_tuple=False)[:,1])
```
This prints:
```
tensor([31,  0, 43, 62], device='mps:0')
tensor([47,  0, 43, 62], device='mps:0')
tensor([15,  0, 43, 62], device='mps:0')
tensor([31,  0, 43, 62], device='mps:0')
tensor([15,  0, 43, 62], device='mps:0')
```
This looks odd since the first element of the tensor should always be ```1``` (not a random number). In the CPU case we have:
```python
mytens = torch.tensor([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                        [1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
                        0., 0., 0., 0., 0., 0., 0., 0., 1., 0.]], device='cpu')

for i in range(5):
    print(torch.nonzero(mytens, as_tuple=False)[:,1])
```
which prints:
```
tensor([ 0,  0, 43, 62])
tensor([ 0,  0, 43, 62])
tensor([ 0,  0, 43, 62])
tensor([ 0,  0, 43, 62])
tensor([ 0,  0, 43, 62])
```
I wonder if this PR solves the problem.

**Comment by kulinseth** [comment]
@alpoge , can you please rebase and fix the lint issues? Otherwise the change looks good to go.

**Review by kulinseth — APPROVED** [review]
Can you please add a Test case to catch this issue in the future.?

**Comment by hvaara** [comment]
Thanks a lot @alpoge! 😃 Please let us know if you need any help with the remaining work.

@pytorchbot label "ciflow/mps"

**Comment by hvaara** [comment]
@mahdiaslanimk I'm not able to repro (tried HEAD, and pytorch@416f5327). Which version of PyTorch did you experience the issue with? Are you able to repro with latest nightly? I can find the commit that fixed the issue if you have a known bad commit sha. If you're still experiencing problems, please consider opening a new issue.

**Comment by alpoge** [comment]
@kulinseth hi extremely sorry about the delay!! i'm on this soon... and also the matmul bugs... :D
